### ISSUE TITLE
Range in characters is sub-optimal for error reporting to humans

### ISSUE THREAD

**Issue by MikeRalphson** [eemeli/yaml#67]
As YAML is line-delimited (can't be strung together minified like JSON) it makes at least some sense to be able to report errors (both syntactically and semantically) as a line number and column number pair.

Is it simple to add a helper routine to achieve this? Happy to PR with a bit of pointing in the right direction.

My end use-case is to turn JSON pointer references to schema violations back into line number, column number pairs.

**Comment by eemeli** [comment]
That sounds like a good idea, and a PR would indeed be welcome. This may need to expose two different APIs, e.g. something like `charPosToLineCol(pos, str): { line: number, col: number }` for the actual implementation (you may be able to find something already on npm) and a wrapper convenience method added to the errors (which due to this may need a common ancestor).

Things to be aware of:
- A `Document` or its parsed nodes do not directly contain any reference to the source string, unless `keepCstNodes` is [enabled](https://eemeli.org/yaml/#options).
- Parsed AST nodes contain a `range: [start, end]` property; nodes not generated by parsing obviously will not.
- CST nodes always have a `range: { start: number, end: number }` property
- Errors always contain a `source` reference to their CST node, which in turn do contain a reference to the source string as `context.src`
- When parsing, if the input contains `\r\n` line terminators, those are converted to `\n` terminators before processing. The string at `context.src` will only have `\n` line terminators. The CST parser's [ability](https://eemeli.org/yaml/#cst-setorigranges-bool) to add `origRange` pointers to the real original source isn't (currently) available when using the library via e.g. `YAML.parse()` or `YAML.parseDocument()`, and should not be strictly required for this.

**Comment by MikeRalphson** [comment]
Ok thanks. I'm making some progress, able to build up a list of line offsets at the same point the `\r`s are removed in the CST parser which should allow a new `charPosToLineCol` function to be performant.

Will PR as soon as I have a tested solution. Please consider me working on this issue.

**Comment by eemeli** [comment]
Correct. The parser doesn't otherwise need to track newlines, so we should not require that it does so every time. However, as you may get multiple errors from a single input document, it may well make sense to cache the positions of newlines when they're first determined.

**Comment by MikeRalphson** [comment]
Thanks. Will put the calculation behind an option.

**Comment by MikeRalphson** [comment]
Perfect, thanks.

**Comment by MikeRalphson** [comment]
```
mike:~/nodejs/yaml$ git submodule update --init
Submodule '__tests__/yaml-test-suite' (https://github.com/eemeli/yaml-test-suite.git) registered for path '__tests__/yaml-test-suite'
Submodule 'docs' (https://github.com/eemeli/yaml-docs.git) registered for path 'docs'
Cloning into '/home/mike/nodejs/yaml/docs'...
error: Server does not allow request for unadvertised object 9d0dddc79fe65d64b7c17c23fa144165c2e4d262
Fetched in submodule path '__tests__/yaml-test-suite', but it did not contain 9d0dddc79fe65d64b7c17c23fa144165c2e4d262. Direct fetching of that commit failed.
```

**Comment by eemeli** [comment]
I don't know what's up with that error. That commit [definitely exists](https://github.com/eemeli/yaml-test-suite/tree/9d0dddc79fe65d64b7c17c23fa144165c2e4d262).

**Comment by MikeRalphson** [comment]
Ignore that, I had a clone of YAML-test-suite in that directory I forgot to remove.

**Comment by MikeRalphson** [comment]
> should add developer guidelines with that info.

Let me know if you want a PR for this, and if so, against `README.md` or a new `CONTRIBUTING.md`?
